sdk-development-guidelines.md

```diff
@@ -0,0 +1,200 @@
+
+#### Expose minimal Surface area
+* Keep everything private unless it needs to be public. Every public class, method, property & resource MUST have a strong reason for being public
+* MAY keep internal source code in a `/Private` folder to allow tools to more easily ignore those files
+* MUST hide classes that need to be public for technical reasons but are not intended for SDK clients from documentation
+* SHOULD declare classes as final to avoid unexpected high-jacking of SDK behavior by subclasses
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * SHOULD use `RestrictTo` annotation on classes that are internal to the library
+
+#### Best Practices
+* Simple configuration keys MAY be settable in application's manifest file (Android) or Info.plist (iOS) or passed directly to an API configuration function:
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * the manifest file is read in a [(fake) ContentProvider](https://firebase.googleblog.com/2016/12/how-does-firebase-initialize-on-android.html) and be automatically initialized (if that fits with your library design)
+    * SHOULD NOT require manual initialization such as by overriding the `Application` class or depending on a host app trigger
+    * Complex Configurations MUST use the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) with [value objects](https://en.wikipedia.org/wiki/Value_object) for configuration values
+        * Exception: When using Kotlin [named arguments](https://kotlinlang.org/docs/reference/functions.html#named-arguments) and [default arguments](https://kotlinlang.org/docs/reference/functions.html#default-arguments)
+    * For better testability of your library it is RECOMMENDED to use [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection)
+* For iOS <img src="images/ios.png" alt="drawing" width="15"/>:
+    * Take the https://swift.org/documentation/api-design-guidelines/ into account when designing your public API interface
+
+### Testing
+
+* All library (sub-)projects MUST generate coverage test reports (e.g. jacoco, xcode/slather)
+* Unit test coverage MUST be at least 80% and SHOULD be above 85%
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * Unit tests MUST NOT rely on reflection to expose otherwise private/internal classes or members
+    * Tests SHOULD use JUnit, and MAY use helper libraries such as Mockito or Robolectric
+* For iOS <img src="images/ios.png" alt="drawing" width="15"/>:
+    * Tests SHOULD use XCTest, and MAY use helper libraries such as OCMock or OHHTTPStubs
+    * Tests MAY be BDD style using Kiwi for Obj-C or Quick/Nimble for Swift
+
+### Documentation
+
+* Public API MUST be annotated with a standard documenting style e.g. Doxygen, Appledoc, Javadoc or KDoc
```

Comment:
Please remove Doxygen from this list - I don't want to encourage use of that

Reply:
removed